finality-provider/service/fp_instance.go

```diff
@@ -1016,3 +1016,83 @@ func (fp *FinalityProviderInstance) GetFinalityProviderSlashedOrJailedWithRetry(
 
 	return slashed, jailed, nil
 }
+
+// CommitPubRandTiming - helper struct used to capture times for benchmark
+type CommitPubRandTiming struct {
+	GetPubRandListTime      time.Duration
+	AddPubRandProofListTime time.Duration
+	CommitPubRandListTime   time.Duration
+}
+
+// HelperCommitPubRand used for benchmark
+func (fp *FinalityProviderInstance) HelperCommitPubRand(tipHeight uint64) (*types.TxResponse, *CommitPubRandTiming, error) {
+	lastCommittedHeight, err := fp.GetLastCommittedHeight()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var startHeight uint64
+	switch {
+	case lastCommittedHeight == uint64(0):
+		// the finality-provider has never submitted public rand before
+		startHeight = tipHeight + 1
+	case lastCommittedHeight < uint64(fp.cfg.MinRandHeightGap)+tipHeight:
+		// (should not use subtraction because they are in the type of uint64)
+		// we are running out of the randomness
+		startHeight = lastCommittedHeight + 1
+	default:
+		fp.logger.Debug(
+			"the finality-provider has sufficient public randomness, skip committing more",
+			zap.String("pk", fp.GetBtcPkHex()),
+			zap.Uint64("block_height", tipHeight),
+			zap.Uint64("last_committed_height", lastCommittedHeight),
+		)
+		return nil, nil, nil
+	}
+
+	return fp.commitPubRandPairsWithTiming(startHeight)
+}
+
+func (fp *FinalityProviderInstance) commitPubRandPairsWithTiming(startHeight uint64) (*types.TxResponse, *CommitPubRandTiming, error) {
+	timing := &CommitPubRandTiming{}
+
+	activationBlkHeight, err := fp.cc.QueryFinalityActivationBlockHeight()
+	if err != nil {
+		return nil, timing, err
+	}
+
+	startHeight = max(startHeight, activationBlkHeight)
+
+	// Measure getPubRandList
+	pubRandListStart := time.Now()
+	pubRandList, err := fp.getPubRandList(startHeight, fp.cfg.NumPubRand)
+	if err != nil {
+		return nil, timing, fmt.Errorf("failed to generate randomness: %w", err)
+	}
+	timing.GetPubRandListTime = time.Since(pubRandListStart)
+
+	numPubRand := uint64(len(pubRandList))
+	commitment, proofList := types.GetPubRandCommitAndProofs(pubRandList)
+
+	// Measure addPubRandProofList
+	addProofStart := time.Now()
+	if err := fp.pubRandState.addPubRandProofList(pubRandList, proofList); err != nil {
+		return nil, timing, fmt.Errorf("failed to save public randomness to DB: %w", err)
+	}
+	timing.AddPubRandProofListTime = time.Since(addProofStart)
+
+	// Measure CommitPubRandList
```

Comment:
Maybe we can just measure the total time of `commitPubRandPairs` and this part can be the total time minus the above